File: 243-shortest-word-distance/243.py
class Solution:
    def shortestDistance(self, wordsDict: List[str], word1: str, word2: str) -> int:
        table = {}
        for i in range(len(wordsDict)):
            word = wordsDict[i]
            if word in table:
                table[word].append(i)
            else:
                table[word] = [i]

        pos1 = table[word1]
        pos2 = table[word2]

        diff = [math.inf] * len(pos1)
        for i in range(len(pos1)):
            for j in range(len(pos2)):
                diff[i] = min(diff[i], abs(pos2[j]-pos1[i]))

        return min(diff)

File: 243-shortest-word-distance/243.cpp
class Solution {
public:
    int shortestDistance(vector<string>& wordsDict, string word1, string word2) {
        vector<int> word1_index, word2_index;
        for (int i = 0; i < wordsDict.size(); ++i) {
            if (wordsDict[i] == word1) {
                word1_index.push_back(i);
            } else if (wordsDict[i] == word2) {
                word2_index.push_back(i);
            }
        }
        int distance = INT_MAX;
        for (int w1 : word1_index) {
            for (int w2 : word2_index) {
                int d = abs(w1 - w2);
                if (d == 1) {
                    return 1;
                }
                distance = min(distance, d);
            }
        }
        return distance;
    }
};
